### ISSUE TITLE
Are location assignments correct in SPIRVReflection.h getShaderOutputs()?

### ISSUE THREAD

**Issue by mbechard** [KhronosGroup/MoltenVK#1501]
I'm having issues with incorrect location/size specification resulting in invalid MSL code.
The code that assigns these locations seems incorrect. In my case I have a GLSL vertex output structure such as:
```
out Vertex
{
    vec4 color;
    mat3 tangentToWorld;
    vec3 worldSpacePos;
    vec2 texCoord0;
    int cameraIndex;
} iVert;
```

The structure has a base location=0 in the SPIR-V. When parsing the reflection it assigns locations for the shader outputs as:
color = 0
tangentToWorld has columns == 3, so it gets 1, 2, 3
worldSpacePos = 2
texCoord0 = 3
cameraIndex = 4

Shouldn't the loop be keeping track of the highest location it's used at the end of each iteration, since on loops where columns or array size > 1, it'll have consumed multiple locations?

**Comment by billhollings** [comment]
Thanks for posting. Yes, the number is incorrect when flattening complex structures.

Can you post this to [SPIRV-Cross issues list](https://github.com/KhronosGroup/SPIRV-Cross/issues), where it can be dealt with, please?

**Comment by mbechard** [comment]
Ok, I was going to do a pull request. Why in SPIRV-Cross though? The file is in this repo.

**Comment by billhollings** [comment]
> Why in SPIRV-Cross though? The file is in this repo.

I assumed the issue was with the main loop that is pulling out of SPIRV-Cross.

Are you referring to the short loop at the bottom of that function, that assigns locations that are not retrieved from SPIRV-Cross?

```C++
// Assign locations to outputs that don't have one.
uint32_t loc = -1;
for (SPIRVShaderOutput& out : outputs) {
	if (out.location == uint32_t(-1)) { out.location = loc + 1; }
	loc = out.location;
}
```

If so...yes...you are probably correct. This should probably be rewritten so that the `loc` is incremented not just by `1` as here, but by the matrix column count or array size of the current output component.

If you have a PR, then please submit it and we can review. Thanks.

**Comment by mbechard** [comment]
That spot is another area that I think is incorrect, but for the first PR I was looking at the earlier loop when it's handling the
`if (type->basetype == SPIRV_CROSS_NAMESPACE::SPIRType::Struct) {`

clause. I'll submit a PR. Thanks

**Comment by mbechard** [comment]
I'm not sure if the loop at the end is needed actually, since legal Vulkan SPIR-V should have had locations either explicitly or implicitly known during the initial loops:
```
An OpVariable with a structure type that is not a block must be decorated with a Location.

When an OpVariable with a structure type (either block or non-block) is decorated with a Location, the members in the structure type must not be decorated with a Location. The OpVariable’s members are assigned consecutive locations in declaration order, starting from the first member, which is assigned the location decoration from the OpVariable.

When a block-type OpVariable is declared without a Location decoration, each member in its structure type must be decorated with a Location. Types nested deeper than the top-level members must not have Location decorations.
```

**Comment by billhollings** [comment]
Fixed via #1502.
